pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -476,6 +453,49 @@ impl WalreceiverState {
         None
     }
 
+    /// Selects the best possible cancidate out of all data, stored from etcd updates about the safeekeepers.
+    /// Optionally, omits the node given, to support the connections witch logic.
+    ///
+    /// The candidate that is chosen:
+    /// * has smallest possible connection attempts from pageserver to safekeeper node (reset every time the WAL replication feedback is sent)
+    /// * has biggest data Lsn among the ones that are left
+    ///
+    /// We do not evict timeline data received from etcd, so there's a chance to select the node with good information that did not send etcd updates for a long time.
```

Comment:
Well, we do evict in `cleanup_old_candidates`. Another problem is possible: choosing lagging node which has 0 attempts counter while there might be more advanced one with non-zero counter (thought it might be already be healthy).